creator-node/src/snapbackSM/peerSetManager.js

```diff
@@ -0,0 +1,221 @@
+      }
+    }
+
+    return unhealthyPeers
+  }
+
+  /**
+   * Retrieve list of all users which have this node as replica (primary or secondary) from discovery node
+   * Or retrieve primary users only if connected to old discprov
+   *
+   * Also handles backwards compatibility of getAllNodeUsers() and getNodePrimaryUsers()
+   * This only works if both functions have a consistent return format
+   */
+  async getNodeUsers () {
+    let fetchUsersSuccess = false
+    let nodeUsers
+
+    let firstFetchError = null
+    try {
+      // Retrieves users from route `v1/full/users/content_node/all`
+      nodeUsers = await this.getAllNodeUsers()
+      fetchUsersSuccess = true
+    } catch (e) {
+      firstFetchError = e
+    }
+
+    if (!fetchUsersSuccess) {
+      try {
+        // Retrieves users from route `users/creator_node`
+        nodeUsers = await this.getNodePrimaryUsers()
+      } catch (secondFetchError) {
+        throw new Error(`getAllNodeUsers() Error: ${firstFetchError.toString()}\n\ngetNodePrimaryUsers() Error: ${secondFetchError.toString()}`)
+      }
+    }
+
+    // Ensure every object in response array contains all required fields
+    nodeUsers.forEach(nodeUser => {
+      const requiredFields = ['user_id', 'wallet', 'primary', 'secondary1', 'secondary2']
+      const responseFields = Object.keys(nodeUser)
+      const allRequiredFieldsPresent = requiredFields.every(requiredField => responseFields.includes(requiredField))
+      if (!allRequiredFieldsPresent) {
+        throw new Error('getNodeUsers() Error: Unexpected response format during getAllNodeUsers() or getNodePrimaryUsers() call')
+      }
+    })
+
+    return nodeUsers
+  }
+
+  /**
+   * Retrieve users with this node as replica (primary or secondary)
+   *  - Makes single request to discovery node to retrieve all users
+   *
+   * @notice This function depends on a new discprov route and cannot be consumed until every discprov exposes that route
+   *    It will throw if the route doesn't exist
+   * @returns {Object[]} array of objects
+   *  - Each object has schema { primary, secondary1, secondary2, user_id, wallet }
+   */
+  async getAllNodeUsers () {
+    // Fetch discovery node currently connected to libs as this can change
+    if (!this.discoveryProviderEndpoint) {
+      throw new Error('No discovery provider currently selected, exiting')
+    }
+
+    // Request all users that have this node as a replica (either primary or secondary)
+    const requestParams = {
+      method: 'get',
+      baseURL: this.discoveryProviderEndpoint,
+      url: `v1/full/users/content_node/all`,
+      params: {
+        creator_node_endpoint: this.creatorNodeEndpoint
+      }
+    }
+
+    // Will throw error on non-200 response
+    let allNodeUsers
+    try {
+      const resp = await axios(requestParams)
+      allNodeUsers = resp.data.data
+    } catch (e) {
+      throw new Error(`getAllNodeUsers() Error: ${e.toString()}`)
```

Comment:
we don't typically do `${e.toString()}`. curious about the switch - is this better? down to start using if so

Reply:
if for any reason there is no `e.message` property, I chose to use the inherit `toString()` method to stringify the err object itself. sometimes I think stringifying the error message isn't structured well, but I rather a chunky looking string than no error message

Reply:
discussed offline